File: Chap5-ModelViewArchitectureAdvanced/5-16DelegatesChallengeDemo/stardelegate.cpp
#include "stardelegate.h"
#include "stareditor.h"
#include <QPainter>

StarDelegate::StarDelegate(QObject *parent) : QStyledItemDelegate(parent)
{
    poly << QPoint(0, 85) << QPoint(75, 75)
         << QPoint(100, 10) << QPoint(125, 75)
         << QPoint(200, 85) << QPoint(150, 125)
         << QPoint(160, 190) << QPoint(100, 150)
         << QPoint(40, 190) << QPoint(50, 125)
         << QPoint(0, 85);
}

void StarDelegate::paint(QPainter *painter, const QStyleOptionViewItem &option, const QModelIndex &index) const
{
    if(index.column() ==3){

        QRect rect = option.rect.adjusted(10,10,-10,-10);

        int starNumber = index.data().toInt();


        painter->save();

        painter->setRenderHint(QPainter::Antialiasing, true);
        painter->setPen(Qt::NoPen);
        painter->setBrush(QBrush(Qt::black));

        painter->translate(rect.x(),rect.y());

        painter->scale(0.1,0.1);

        for(int i = 0; i < starNumber ; i ++){
            painter->drawPolygon(poly);
            painter->translate(220,0);
        }

        painter->restore();

    }

}

QSize StarDelegate::sizeHint(const QStyleOptionViewItem &option, const QModelIndex &index) const
{
    return QStyledItemDelegate::sizeHint(option, index);
}

QWidget *StarDelegate::createEditor(QWidget *parent, const QStyleOptionViewItem &option,
                                    const QModelIndex &index) const
{
     if(index.column() ==3){
         int starRating = index.data().toInt();
         StarEditor * editor = new StarEditor(parent);
         editor->setStarRating(starRating);
         connect(editor,&StarEditor::editingFinished,this,&StarDelegate::commitAndCloseEditor);
         return editor;
     }else{
         return QStyledItemDelegate::createEditor(parent,option,index);
     }
}

void StarDelegate::setEditorData(QWidget *editor, const QModelIndex &index) const
{
     if(index.column() ==3){
         int starRating = index.data().toInt();
         StarEditor *starEditor = qobject_cast<StarEditor *>(editor);
         starEditor->setStarRating(starRating);
     }else{
         QStyledItemDelegate::setEditorData(editor,index);
     }
}

void StarDelegate::setModelData(QWidget *editor, QAbstractItemModel *model, const QModelIndex &index) const
{
     if(index.column() ==3){
         StarEditor *starEditor = qobject_cast<StarEditor *>(editor);
         model->setData(index,QVariant::fromValue(starEditor->getStarRating()),
                        Qt::EditRole);
     }else{
         QStyledItemDelegate::setModelData(editor,model,index);
     }

}

void StarDelegate::commitAndCloseEditor()
{
    StarEditor *editor = qobject_cast<StarEditor *>(sender());
    emit commitData(editor);
    emit closeEditor(editor);
}

File: Chap5-ModelViewArchitectureAdvanced/5-15CustomStarDelegateDemo/stardelegate.py
from PySide6.QtWidgets import QStyledItemDelegate, QWidget
from PySide6.QtCore import Qt, QSize, QPoint, Slot
from PySide6.QtGui import QPainter, QPolygon, QBrush
from stareditor import StarEditor

class StarDelegate(QStyledItemDelegate):
    def __init__(self, parent=None):
        super().__init__(parent)
        
        # Create star polygon shape
        self.poly = QPolygon()
        self.poly << QPoint(0, 85) << QPoint(75, 75) \
                 << QPoint(100, 10) << QPoint(125, 75) \
                 << QPoint(200, 85) << QPoint(150, 125) \
                 << QPoint(160, 190) << QPoint(100, 150) \
                 << QPoint(40, 190) << QPoint(50, 125) \
                 << QPoint(0, 85)
    
    def paint(self, painter, option, index):
        """Custom paint method to draw stars"""
        if index.column() == 2:
            rect = option.rect.adjusted(10, 10, -10, -10)
            star_number = index.data()
            
            # Convert from QVariant if necessary
            if isinstance(star_number, str):
                star_number = int(star_number)
            
            painter.save()
            
            # Enable antialiasing for smoother drawing
            painter.setRenderHint(QPainter.RenderHint.Antialiasing, True)
            painter.setPen(Qt.PenStyle.NoPen)
            painter.setBrush(QBrush(Qt.GlobalColor.black))
            
            # Move painter for star drawing
            painter.translate(rect.x(), rect.y())
            painter.scale(0.1, 0.1)
            
            # Draw the stars based on rating
            for i in range(star_number):
                painter.drawPolygon(self.poly)
                painter.translate(220, 0)
            
            painter.restore()
        else:
            # For other columns, use default painting
            super().paint(painter, option, index)
    
    def sizeHint(self, option, index):
        """Return size hint for the item"""
        return super().sizeHint(option, index)
    
    def createEditor(self, parent, option, index):
        """Create a custom editor for star ratings"""
        if index.column() == 2:
            star_rating = index.data()
            
            # Convert from QVariant if necessary
            if isinstance(star_rating, str):
                star_rating = int(star_rating)
            
            # Create and configure the star editor
            editor = StarEditor(parent)
            editor.setStarRating(star_rating)
            
            # Connect to the editing finished signal
            editor.editingFinished.connect(self.commitAndCloseEditor)
            
            return editor
        else:
            # For other columns, use default editor
            return super().createEditor(parent, option, index)
    
    def setEditorData(self, editor, index):
        """Set the editor data from the model"""
        if index.column() == 2:
            star_rating = index.data()
            
            # Convert from QVariant if necessary
            if isinstance(star_rating, str):
                star_rating = int(star_rating)
            
            # Set the star rating in the editor
            star_editor = editor
            star_editor.setStarRating(star_rating)
        else:
            # For other columns, use default behavior
            super().setEditorData(editor, index)
    
    def setModelData(self, editor, model, index):
        """Update the model with data from the editor"""
        if index.column() == 2:
            star_editor = editor
            # Set the data in the model - directly use Python int
            model.setData(index, star_editor.getStarRating(), Qt.ItemDataRole.EditRole)
        else:
            # For other columns, use default behavior
            super().setModelData(editor, model, index)
    
    @Slot()
    def commitAndCloseEditor(self):
        """Commit data and close the editor"""
        editor = self.sender()
        self.commitData.emit(editor)
        self.closeEditor.emit(editor)